src/Native/System.Native/pal_interfaceaddresses.cpp

```diff
@@ -0,0 +1,147 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#include "pal_config.h"
+#include "pal_interfaceaddresses.h"
+#include "pal_maphardwaretype.h"
+#include "pal_utilities.h"
+
+#include <assert.h>
+#include <ifaddrs.h>
+#include <net/if.h>
+#include <netinet/in.h>
+#include <sys/socket.h>
+#include <sys/sysctl.h>
+
+static_assert(HAVE_AF_PACKET || HAVE_AF_LINK, "System must have AF_PACKET or AF_LINK.");
+#if HAVE_AF_PACKET
+#include <linux/if_packet.h>
+#elif HAVE_AF_LINK
+#include <net/if_dl.h>
+#include <net/if_types.h>
+#endif
+
+#if HAVE_RT_MSGHDR
+#include <net/route.h>
+#endif
+
+extern "C" int32_t EnumerateInterfaceAddresses(IPv4AddressFound onIpv4Found,
+                                            IPv6AddressFound onIpv6Found,
+                                            LinkLayerAddressFound onLinkLayerFound)
+{
+    ifaddrs *headAddr, *current;
+    if (getifaddrs(&headAddr) == -1)
+    {
+        return -1;
+    }
+
+    for (current = headAddr; current != nullptr; current = current->ifa_next)
+    {
+        uint32_t interfaceIndex = if_nametoindex(current->ifa_name);
+        int family = current->ifa_addr->sa_family;
+        if (family == AF_INET)
+        {
+            // IP Address
+            IpAddressInfo iai = {};
```

Comment:
Super nit: instead of the empty initialization followed by piecewise member assignments, this could be:

```
IpAddressInfo iai = {
    .InterfaceIndex = interfaceIndex,
    .NumAddressBytes = NUM_BYTES_IN_IPV4_ADDRESS,
};

sockaddr_in* sain = reinterpret_cast<sockaddr_in*>(current->ifa_addr);
*reinterpret_cast<uint32_t*>(iai.AddressBytes) = sain->sin_addr.s_addr;
```